---
title: "results_cleaning_and_viz"
author: "MV"
date: "23/01/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(readr)
library(tidyr)
library(data.table)
library(ggplot2)
library(wesanderson)
library(plotly)
library(colorspace)
library(cowplot)
```

```{r}
data_path<-"/XremovedX" # to use in Rstudio 
#data_path<-"XremovedX"
#data_path<-"/projects/XremovedX" #to use in terminal
```


```{r}
# load data annotated in script 4_
univariate_results<-"test_result_w11_r_UKB9M/" 

with_trait<- fread(paste0(data_path, "results/", univariate_results,"/metaCCA_snps_all_UKB_annotated_w_traits.tsv"))
with_trait<- fread(paste0(data_path, "results/", univariate_results,"/metaCCA_snps_all_UKB_annotated_w_traits_addedProxies.tsv"))
```

```{r}

dat<- with_trait %>% 
          #filter(!is.na(trait_summary)) %>% 
          mutate(pval_truncated= ifelse(pval < 1e-15, 1e-15, pval)) %>% 
          mutate(log10pval_trunc= as.integer(-log10(pval_truncated))) %>% 
          mutate(trait_summary=ifelse(trait_summary == 0, "x_miscellaneous", trait_summary)) %>% 
          mutate(full_trait=MAPPED_TRAIT) %>% 
          select(gene, SNPS, r_1, log10pval_trunc, full_trait, proxy_rsq, 'STUDY ACCESSION', trait_summary) 

dat %>% select(trait_summary) %>% plyr::count() %>% arrange(freq)

#write_csv(dat, "~/Documents/My Tableau Repository/Datasources/metaCCA/test_input.csv")
```




GENES
```{r}
# load dat from gene analysis where we got traits by  genes, and the got traits only for snps that has associations

#dat<-read_tsv(paste0(data_path, "gwas_catalog/test_results.tsv"))
#metacca_genes <- dat %>%
#                mutate(pval_truncated= ifelse(pval < 1e-15, 1e-15, pval)) %>% 
#                mutate(log10pval_trunc= as.integer(-log10(pval_truncated))) %>% 
#                mutate(gene_snp = paste0(gene,"_", SNPS)) %>% 
#                select(r_1, log10pval_trunc, MAPPED_TRAIT, gene_snp, trait_summary) %>% distinct


```








# scatter plot
```{r}
ggplot(data = dat) + 
  geom_point(mapping = aes(x = r_1, y = log10pval_trunc))
```

# SET COLOUR PALETTES
```{r}

pal1 <- wes_palette("Rushmore1")[c(1,2,3,5)]
pal2 <- wes_palette("Zissou1")
pal3 <- wes_palette("FantasticFox1")
pal4 <- wes_palette("Darjeeling1")


set.seed(009) # just to make it reproducible
megapal<-sample(c(pal1,pal2,pal3,pal4,pal2))

# manual 14 cols
megapal <- c("#78B7C5", "#EABE94", "#EBCC2A", "#FF0000", "#3B9AB2" , "#B40F20", "#0B775E",
             "#F2300F", "#DD8D29",  "#5BBCD6", "#E2D200" , "#F98400", "#46ACC8", "#00A08A",
             "#F98400", "#46ACC8", "#00A08A")

hue_pal<-c("#ff823f", "#284bb0", "#99b000", "#ff9ef3", "#5f9e00", "#ff5f87", "#02cd7e", "#ac2100", "#02ffe8", "#ffa879", "#2cb7ff", "#e7fd86", "#007d5c", "#ffd871", "#919976", "black")

library(scales)
show_col(megapal)

```


# Box plot
```{r}

#trait_order<- dat %>%  select(trait_summary) %>% plyr::count() %>% arrange(desc(freq)) %>% pull(trait_summary)
#dat$trait_summary <- factor(dat$trait_summary,levels = trait_order, ordered = TRUE)
dat_full<-dat
dat<- dat_full %>% filter(proxy_rsq >= 0.8)
dim(dat)

p<-ggplot(data = dat, mapping = aes(x = trait_summary, y = r_1, fill= trait_summary)) + 
  geom_boxplot(varwidth = F)+
  scale_fill_manual(values = megapal)+
  theme_minimal_hgrid(10, rel_small = 1) +
  #ylim(0,0.21)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  labs(color = "Trait groups",
       y = "Canonical Correlation (CC) value (r_1)", x = "",
       #title="Boxplots of CC values for SNP with trait annotation in GWAS catalog",
       title="metaCCA CC values grouped into trait categories, extracted from GWAS catalog",
       subtitle = paste0("Based on UKB IEU 7 traits with 9M genotype-phenotype matrix \nNumber of SNPs displayed: ", dim(dat)[1], ": only real SNPs + proxies r2 > 0.80"))


ggsave("~/Desktop/boxplot.png",
       plot=p, scale=1, 
       width=17, height=20,
       units=c("cm"), dpi=200, limitsize=F)
```


# Dot plot
```{r}

p<-ggplot(data = dat,
          mapping = aes(x = trait_summary, label=full_trait, label2=gene,
                        y = r_1, color= trait_summary, 
                        size=log10pval_trunc )) + 
  geom_jitter(alpha=0.2, position = position_jitter(width = 0.3))+
  ylim(0,0.09)+
  scale_color_manual(values=darken(megapal, 0.08))+
  scale_fill_manual(values = megapal) +
  theme_minimal_hgrid(10, rel_small = 1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position="top",
        legend.box.background = element_blank())+
  labs(color = "Trait groups",
       size = "-log10pval",
       y = "Canonical Correlation value (r_1)", x = "",
       title="CC values for SNPs annotated with GWAS catalog ",
       subtitle = "Based on UKB IEU 7 traits with 9M genotype-phenotype matrix")+
  guides(alpha = FALSE,colour=FALSE)
         #color= guide_legend(order = 1, override.aes = list(alpha = 1)))

ggsave("~/Desktop/test.png",
       plot=p, scale=1, 
       width=17, height=20,
       units=c("cm"), dpi=200, limitsize=F)

ply <- ggplotly(p)

```




# Barchart


```{r}
data<- with_trait %>%
            mutate(trait_summary=ifelse(is.na(trait_summary), "Not in GWAScat", trait_summary)) %>% 
            mutate(trait_summary=ifelse(trait_summary == 0, "misc", trait_summary)) %>% 
            filter(pval < 10e-8) %>% 
            filter(r_1 > 0.02) %>% 
            # create r_1 bins
            mutate(r_1_bins=(ifelse(r_1 > 0.15 , "> 0.15", 
                             ifelse(r_1 > 0.1 , "> 0.1", 
                             ifelse(r_1 > 0.07 , "> 0.07",
                             ifelse(r_1 > 0.06 , "> 0.06",
                             ifelse(r_1 > 0.05 , "> 0.05",
                             ifelse(r_1 > 0.045 , "> 0.045",
                             ifelse(r_1 > 0.04 , "> 0.04",
                             ifelse(r_1 > 0.035 , "> 0.035",
                             ifelse(r_1 > 0.03 , "> 0.03",
                             ifelse(r_1 > 0.027 , "> 0.027",
                             ifelse(r_1 > 0.024 , "> 0.024",
                             ifelse(r_1 > 0.022 , "> 0.022",
                             ifelse(r_1 > 0.02 , "> 0.02")))))))))))))))
            
data %>% select(r_1_bins) %>% plyr::count()
```

```{r}

trait_order<- data %>%  select(trait_summary) %>% plyr::count() %>% arrange(desc(freq)) %>% pull(trait_summary)
data$trait_summary <- factor(data$trait_summary,levels = trait_order, ordered = TRUE)

# if want to drop Not in GWAS
data<-filter(data, trait_summary != "Not in GWAScat")

p<- ggplot(data = data, mapping = aes(x = r_1_bins, fill= trait_summary)) + 
    geom_bar()+
    #scale_fill_manual(values=c("black", megapal))+ #with "not in GWAS"
    scale_fill_manual(values=c(megapal))+
    labs(fill = "Trait groups",
         y = "Count (included only pval < 10e-8 and r_1 > 0.02)", x = "r_1 (CC value) arbitary groups",
         title="Count of traits per CCvalue group, by trait group in GWAS catalog")+
    theme_minimal()
    
    
ply <- ggplotly(p)

```

