---
title: "meta CCA toy examples walkthrough"
output: html_notebook
---


```{r}
library(metaCCA)
data( package = 'metaCCA' )
```
*metaCCA* operates on three pieces of the full data covariance matrix:

- S_XY of univariate genotype-phenotype association results
- S_XX of genotype-genotype correlations
- S_YY of phenotype-phenotype correlations.

S_XX is estimated from a reference database matching the study population, e.g. the 1000 Genomes.
S_YY is estimated from S_XY.

```{r}

print(N1) # individulas in study 1
dim(S_XX_study1) # XX reference
dim(S_XY_study1) # summary stats (test) = associaion results
dim(S_XY_full_study1)

head(S_XX_study1)
head((S_XY_study1))
```

- traitID b - univariate regression coefficients; 
- traitID se - corresponding standard errors;


## Workflow
### Estimation of phenotypic correlation structure S_YY
Correlations between traits (S_YY) are estimated from univariate summary statistics S_XY.
Here, we will estimate correlations between 10 traits using *estimateSyy* function.
In each case, we will use summary statistics of 1000 SNPs. However, in practice, summary statistics of at least one chromosome should be used in order to ensure good quality
of SY_Y estimate. *estimateSyy* can be used no matter if the univariate analysis has been performed on standardised data (meaning that the genotypes were standardised before regression coefficients and standard errors were computed) or non-standardised data.



```{r}
# Estimating phenotypic correlation structure of study 1 
S_YY_study1 = estimateSyy( S_XY = S_XY_full_study1 )
# Estimating phenotypic correlation structure of study 2 
S_YY_study2 = estimateSyy( S_XY = S_XY_full_study2 )
```

```{r}
#S_YY containing correlations between traits given as input
dim(S_YY_study1)
print( head(S_YY_study1[,1:6]), digits = 3 ) 
```

### Genotype-phenotype association analysis The
- *metaCcaGp* - runs the analysis according to metaCCA algorithm;
- *metaCcaPlusGp* - runs the analysis according to a variant of metaCCA, namely metaCCA+, where the full covariance matrix is shrunk beyond the level guarantee- ing its positive semidefinite property. (supposedly makes it more robust, ie less false positives)

By default, *metaCcaGp* and *metaCcaPlusGp* perform single-SNP-multi-trait analysis, where each given SNP is analysed in turn against all given phenotypic variables.

#### Single-SNP-multi-trait analysis

- *nr_studies* - number of studies analysed.
- *S_XY* - a list of data frames (one for each study) with univariate summary statistics corresponding to SNPs and traits to be included in the analysis

- *std_info* - a vector with indicators (one for each study) if the univariate analysis has been performed on standardised (1) or non-standardised (0) data (most likely the data were not standardised - the genotypes were not standardised before univariate regression coefficients and standard errors were computed - option 0 should be used);
- *S_YY* - a list of matrices (one for each study), estimated using *estimateSyy* function, containing correlations between traits to be included in the analysis;
- *N* - a vector with numbers of individuals in each study.


We will first run the default single-SNP-multi-trait analysis of two studies using provided toy data. Each of 10 SNPs will be tested for an association with the group of 10 traits.
```{r}

# Default single-SNP--multi-trait meta-analysis of 2 studies
# Association analysis according to metaCCA algorithm
metaCCA_res1 = metaCcaGp( nr_studies = 2, 
                          S_XY = list( S_XY_study1, S_XY_study2 ),
                          std_info = c( 0, 0 ),
                          S_YY = list( S_YY_study1, S_YY_study2 ),
                          N = c( N1, N2))



# Association analysis according to metaCCA+ algorithm
metaCCApl_res1 = metaCcaPlusGp( nr_studies = 2,
                                S_XY = list( S_XY_study1, S_XY_study2 ),
                                std_info = c( 0, 0 ),
                                S_YY = list( S_YY_study1, S_YY_study2 ),
                                N = c( N1, N2 ) )

```


```{r}

# me try!
# try with full studies
metaCCA_res1_full = metaCcaGp( nr_studies = 2, 
                          S_XY = list( S_XY_full_study1, S_XY_full_study2 ),
                          std_info = c( 0, 0 ),
                          S_YY = list( S_YY_study1, S_YY_study2 ),
                          N = c( N1, N2))
dim(metaCCA_res1_full)

# try with full, single SNP analysis, but pick  a subles of the full dataset
### CANNOT DO THiS - for this analysis it want so do all SNPS in XY
metaCCA_res2_full = metaCcaGp( nr_studies = 2, 
                          S_XY = list(S_XY_full_study1, S_XY_full_study2),
                          std_info = c( 0, 0 ),
                          S_YY = list( S_YY_study1, S_YY_study2 ),
                          N = c( N1, N2 ),
                          analysis_type = 1,
                          SNP_id = c('rs1', 'r35', 'rs100' ) )


```



The output is a data frame with row names corresponding to SNP IDs. 
The columns contain the following information for each analysed SNP:

- *r_1* - leading canonical correlation value, 
- *-log10(p-val)* - p-value in the -log10 scale, 
- *trait_weights* - trait-wise canonical weights,
- *snp_weights* - SNP-wise canonical weights (only for multi-SNP-multi-trait analysis).

```{r}
dim(metaCCA_res1)
head(metaCCA_res1)
```

```{r}
# Single-SNP--multi-trait meta-analysis of 2 studies
# and one selected SNP metaCCA 
metaCCA_res2 = metaCcaGp( nr_studies = 2, 
                          S_XY = list( S_XY_full_study1, S_XY_full_study2 ),
                          std_info = c( 0, 0 ),
                          S_YY = list( S_YY_study1, S_YY_study2 ),
                          N = c( N1, N2 ),
                          analysis_type = 1,
                          SNP_id = 'rs1' )
# Result of metaCCA 
print( metaCCA_res2, digits = 2 )
##
```


#### Multi-SNP-multi-trait analysis

In order to analyse multiple SNPs jointly, you need to provide the following additional inputs: 

- *analysis_type* - indicator of the analysis type: 2; 
- *SNP_id* - a vector with IDs of SNPs to be analysed jointly;
- *S_XX* - a list of data frames (one for each study) containing correlations between SNPs to be analysed 

Here, we will run the analysis of 5 SNPs with IDs 'rs10', 'rs80', 'rs140', rs170' and 'rs172'. They will be tested jointly for an association with the group of 10 traits

```{r}
# Multi-SNP--multi-trait meta-analysis of 2 studies
metaCCA_res3 = metaCcaGp( nr_studies = 2,
                          S_XY = list( S_XY_full_study1, S_XY_full_study2 ), 
                          std_info = c( 0, 0 ), 
                          S_YY = list( S_YY_study1, S_YY_study2 ),
                          N = c( N1, N2 ),
                          analysis_type = 2,
                          #SNP_id = c( 'rs10', 'rs80', 'rs140', 'rs170', 'rs172'),
                          SNP_id = c( 'rs1', 'rs8', 'rs14', 'rs17', 'rs18'),
                          S_XX = list( S_XX_study1, S_XX_study2 ) )
```

```{r}
dim(metaCCA_res3)
print(metaCCA_res3)
```

